From 6d2b05742a5780f30cf573da4f3e8960e3cd43f5 Mon Sep 17 00:00:00 2001
From: Guy Nisan <gnisan@habana.ai>
Date: Tue, 25 Aug 2020 22:34:49 +0300
Subject: [PATCH] [SW-17562] spi: add otp support

For Gaudi security, it is needed to access OTP for key validation. This
commit adds OTP support at the lower level flash code.

Change-Id: I822ace9016a5c5b3f98999e960448d0945a1eda2
Signed-off-by: Guy Nisan <gnisan@habana.ai>
---

diff --git a/common/spi.c b/common/spi.c
index e5062ff..d8f4b09 100644
--- a/common/spi.c
+++ b/common/spi.c
@@ -34,6 +34,10 @@
 #define CMD_WRITE_ENABLE		0x06
 #define frame_size_bits(b)		((b) << 16)
 
+/* Used for Macronix flashes only. */
+#define SPINOR_OP_ENSO			0xB1
+#define SPINOR_OP_EXSO			0xC1
+
 static unsigned int spi_regs;
 /*
  * SSI device Chip Select (CS0) is combined with GPIO53
@@ -223,6 +227,14 @@
 	regmap_writel(0x00000001, spi_regs + SSIENR);
 }
 
+void spi_copy_otp(void *dst, unsigned int offset, unsigned int size)
+{
+	spi_init();
+	spi_char_cmd_csel(SPINOR_OP_ENSO, NULL);
+	spi_read(dst, offset, size);
+	spi_char_cmd_csel(SPINOR_OP_EXSO, NULL);
+}
+
 void spi_copy(void *dst, unsigned int offset, unsigned int size, char *name)
 {
 	unsigned char sr;
diff --git a/include/spi.h b/include/spi.h
index faa710d..28ea6a6 100644
--- a/include/spi.h
+++ b/include/spi.h
@@ -25,6 +25,7 @@
 #endif
 
 void spi_copy(void *dst, unsigned int offset, unsigned int size, char *name);
+void spi_copy_otp(void *dst, unsigned int offset, unsigned int size);
 u32 get_cs_pad_sel(void);
 
 #endif /* _HL_SPI_H_ */
